RFCs/FS-1142-Extended-numeric-literal.md

```diff
@@ -0,0 +1,137 @@
+- [x] Approved in principle
+- [ ] [Implementation](https://github.com/dotnet/fsharp/pull/17242)
+- [ ] Design Review Meeting(s) with @dsyme and others invitees
+- [Discussion](https://github.com/fsharp/fslang-design/discussions/769)
+
+# Summary
+
+This RFC will allow the following things:
+
+- Underscores in numeric literals after prefix and before suffix like `0x_1` or `1_l` or mixed them up like `0x_1_l`.
+- Hexadecimal, octal, binary and floating point custom numeric literals like `0x1I` or `1.0G`
+
+# Motivation
+
+Make the language more consistent and easier to read. Enhance the custom numeric literals feature by supporting not only integers but also floats.
+
+# Detailed design
+
+1. Underscores can now be placed after prefix (`0x`, `0o`, `0b`) and before suffix (eg. `UL`, `y`, `m`, `f`). These are forbidden [previously](https://github.com/fsharp/fslang-design/blob/main/FSharp-4.1/FS-1005-underscores-in-numeric-literals.md#detailed-design).
+
+    ```fsharp
+    let pi1 = 3_.1415F      // Invalid cannot put underscores adjacent to a decimal point
+    let pi2 = 3._1415F      // Invalid cannot put underscores adjacent to a decimal point
+    let socialSecurityNumber1 = 999_99_9999_L         // OK (Invalid *previously*)
+
+    let x1 = _52              // This is an identifier, not a numeric literal
+    let x2 = 5_2              // OK (decimal literal)
+    let x3 = 52_              // Invalid cannot put underscores at the end of a literal
+    let x4 = 5_______2        // OK (decimal literal)
+
+    let x5 = 0_x52            // Invalid cannot put underscores in the 0x radix prefix
+    let x6 = 0x_52            // OK (Invalid *previously*)
+    let x7 = 0x5_2            // OK (hexadecimal literal)
+    let x8 = 0x52_            // Invalid cannot put underscores at the end of a number
+
+    // In contrast to Java, literals with leading zeros are decimal in F#.
+    let x9 = 0_52             // OK (decimal literal)
+    let x10 = 05_2            // OK (decimal literal)
+    let x11 = 052_            // Invalid cannot put underscores at the end of a number
+
+    // To create an octal literal, prefix it with '0o' similar to hexadezimal literals. The same rules apply:
+    let x12 = 0_o52            // Invalid cannot put underscores in the 0o radix prefix
+    let x13 = 0o_52            // OK (Invalid *previously*)
+    let x14 = 0o5_2            // OK (octal literal)
+    let x15 = 0o52_            // Invalid cannot put underscores at the end of a number
+    ```
+
+2. Allow number prefix (`0x`, `0o`, `0b`) before integer custom numeric literals.
+  
+    ```fsharp
+    let x1 = 0x123I        // big int (291)
+    let x2 = 0o123I        // big int (83)
+    let x3 = 0b1010I       // big int (10)
+    ```
+
+    The number prefix **will not be removed** from the string sended to the `FromString` of the numeric literal module. This may be a break change.
```

Comment:
No, this is not a breaking change, because it was previously not possible to do this.
```suggestion
    The number prefix is retained when passed as argument to `FromString` of the numeric literal module. 
```

Reply:
Although the bignum (`123I`) already known how to deal with the number prefix (`0x123I` is now ok without changing the library), other custom numeric literals may not know that. This might break their function, so I think it is a breaking change.

Reply:
That is, this is not a break change for existing source code, but it is for library.